fem/hyperbolic.hpp

```diff
@@ -222,62 +373,269 @@ class HyperbolicFormIntegrator : public NonlinearFormIntegrator
+    * side = 2:
+    *    ½J(u⁺,x)n - ½λ
+    */
+   void Grad(int side, const Vector &state1, const Vector &state2,
+             const Vector &nor, FaceElementTransformations &Tr,
+             DenseMatrix &grad) const override;
+
+   /**
+    * @brief  Average normal numerical flux over the interval [u⁻, u⁺] in the
+    * second argument of the flux F̂(u⁻,u,x) n
+    * @note The average normal flux F̄ n is required to be implemented in
+    * FluxFunction::ComputeAvgFluxDotN()
+    * @note Systems of equations are treated component-wise
+    *
+    * @param[in] state1 state value (u⁻) of the beginning of the interval
+    * (num_equations)
+    * @param[in] state2 state value (u⁺) of the end of the interval
+    * (num_equations)
+    * @param[in] nor normal vector (not a unit vector) (dim)
+    * @param[in] Tr face element transformation
+    * @param[out] flux ½(F̄(u⁻,u⁺,x)n + F(u⁻,x)n) - ¼λ(u⁺ - u⁻)
+    * @return max(|dF(u⁺,x)/du⁺⋅n|, |dF(u⁻,x)/du⁻⋅n|)
+    */
+   real_t Average(const Vector &state1, const Vector &state2,
+                  const Vector &nor, FaceElementTransformations &Tr,
+                  Vector &flux) const override;
+
+   /**
+    * @brief  Jacobian of average normal numerical flux over the interval
+    * [u⁻, u⁺] in the second argument of the flux F̂(u⁻,u,x) n
+    * @note The average normal flux F̄ n is required to be implemented in
+    * FluxFunction::ComputeAvgFluxDotN() and the Jacobian of flux J n in
+    * FluxFunction::ComputeFluxJacobianDotN()
+    * @note Only the diagonal terms of the J n are considered, i.e., systems
+    * are treated as a set of independent equations
+    *
+    * @param[in] side gradient w.r.t the first (u⁻) or second argument (u⁺)
+    * @param[in] state1 state value (u⁻) of the beginning of the interval
+    * (num_equations)
+    * @param[in] state2 state value (u⁺) of the end of the interval
+    * (num_equations)
+    * @param[in] nor normal vector (not a unit vector) (dim)
+    * @param[in] Tr face element transformation
+    * @param[out] grad Jacobian of F̄(u⁻,u⁺,x) n
+    * side = 1:
+    *    ½(F̄(u⁻,u⁺,x)n - F(u⁻,x)n) / (u⁺ - u⁻) - ½J(u⁻,x)n + ¼λ
+    * side = 2:
+    *    ½(F(u⁺,x)n - F̄(u⁻,u⁺,x)n) / (u⁺ - u⁻) - ¼λ
+    */
+   void AverageGrad(int side, const Vector &state1, const Vector &state2,
+                    const Vector &nor, FaceElementTransformations &Tr,
+                    DenseMatrix &grad) const override;
+
+protected:
 #ifndef MFEM_THREAD_SAFE
-      fluxN1.SetSize(fluxFunction.num_equations);
-      fluxN2.SetSize(fluxFunction.num_equations);
+   mutable Vector fluxN1, fluxN2;
+   mutable DenseMatrix JDotN;
 #endif
-   }
+};
 
+/**
+ * @brief Godunov flux for scalar equations
+ *
+ * Godunov flux for scalar equations, a special case of Engquist-Osher flux,
+ * is defined as follows:
+ *    F̂ n = min F(u)n    for u⁻ ≤ u⁺, u ∈ [u⁻,u⁺]
+ *    F̂ n = max F(u)n    for u⁻ > u⁺, u ∈ [u⁺,u⁻]
+ * @note The implementation assumes monotonous F(u,x) in u
```

Comment:
This is incorrect: it assumes **convex** F(u,x) in u.

Reply:
Sorry for the mix up with Rusanov (due to that copied over description), the assumptions are different. Here, we need just monotonicity, i.e., min/max F(u) over some interval [u-,u+] be min/max(F(u-), F(u+)). So such function can be, for example, convex (not crossing min), concave (not crossing max) or inflex.

Reply:
No, this is the numerical flux that results when Godunov's method is used with the exact Riemann solution for a scalar conservation law with convex flux, with *convex* being used here in the context as defined by [Wikipedia](https://en.wikipedia.org/wiki/Convex_function) (and elsewhere).

Again, consider that Burgers flux is **not** monotonic, yet this is still the Godunov using the **exact Riemann solution** for Burgers equation (characteristic flow to the right or left, a shock moving left or right, or a rarefaction fan).